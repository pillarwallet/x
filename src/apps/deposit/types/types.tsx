export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
