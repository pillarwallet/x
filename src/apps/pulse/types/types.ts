export type TransactionStatusState =
  | 'Starting Transaction'
  | 'Transaction Pending'
  | 'Transaction Complete'
  | 'Transaction Failed';

export type TransactionStep =
  | 'Submitted'
  | 'Pending'
  | 'ResourceLock'
  | 'Completed';

export type StepStatus = 'completed' | 'pending' | 'failed' | 'inactive';

export interface TokenInfo {
  symbol: string;
  name: string;
  logo: string;
  address?: string;
}

export interface SellOffer {
  tokenAmountToReceive: number;
  minimumReceive: number;
}

export interface PayingToken {
  totalUsd: number;
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalRaw: string;
  chainId: number;
  address: string;
}

export interface TokenDetails {
  symbol: string;
  name: string;
  address: string;
  chainId: number;
  amount: string;
  logo: string;
  type: 'BUY_TOKEN' | 'SELL_TOKEN';
}

export interface BuyModeDetails {
  usdAmount: string;
  payingTokens: PayingToken[];
  totalPayingUsd: number;
}

export interface TransactionStatusConfig {
  icon: string;
  containerClasses: string;
  iconClasses: string;
  color: string;
}

export interface ButtonConfig {
  bgColor: string;
  textColor: string;
  borderColor: string;
  label: string;
}

export interface ProgressStepConfig {
  label: string;
  order: number;
}

export interface TransactionStatusProps {
  closeTransactionStatus: () => void;
  userOpHash: string;
  chainId: number;
  gasFee?: string;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  // Externalized polling state
  currentStatus: TransactionStatusState;
  errorDetails: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  blockchainTxHash?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  completedChainId?: number;
  resourceLockChainId?: number;
  resourceLockCompletedAt?: Date;
  isResourceLockFailed?: boolean;
}

export interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status: TransactionStatusState;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  resourceLockCompletedAt?: Date;
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  isResourceLockFailed?: boolean;
}

export interface TransactionInfoProps {
  status: TransactionStatusState;
  userOpHash: string;
  txHash?: string;
  chainId: number;
  gasFee?: string;
  completedAt?: Date;
  // Buy-specific fields
  isBuy?: boolean;
  resourceLockTxHash?: string;
  resourceLockChainId?: number;
  completedTxHash?: string;
  completedChainId?: number;
}

export interface ProgressStepProps {
  step: TransactionStep;
  status: StepStatus;
  label: string;
  isLast?: boolean;
  showLine?: boolean;
  lineStatus?: StepStatus;
  timestamp?: string | React.ReactNode;
}

export interface TransactionErrorBoxProps {
  technicalDetails?: string;
}

export interface UseClickOutsideOptions {
  ref: React.RefObject<HTMLElement>;
  callback: () => void;
  condition: boolean;
}

export interface UseKeyboardNavigationOptions {
  onEscape: () => void;
  onEnter?: () => void;
  enabled?: boolean;
}

export interface TechnicalDetails {
  transactionType: 'BUY' | 'SELL';
  transactionHash: string;
  hashType: 'bidHash' | 'userOpHash';
  chainId: number;
  status: TransactionStatusState;
  timestamp: string;
  accountAddress: string;
  token: TokenDetails | null;
  sellOffer: SellOffer | null;
  buyMode: BuyModeDetails | null;
  transactionHashes: {
    [key: string]: string;
  };
  chains: {
    mainChainId: number;
    resourceLockChainId: string | number;
    completedChainId: string | number;
  };
  timestamps: {
    [key: string]: string;
  };
  error: {
    details: string;
    isResourceLockFailed: boolean;
    failureStep: string;
  };
  gas: {
    fee: string;
  };
  stepStatus: {
    [key: string]: StepStatus;
  };
}
