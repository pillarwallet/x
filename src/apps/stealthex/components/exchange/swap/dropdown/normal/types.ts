
import type React from 'react';
import { StructuredCurrency } from '../../../../../type';
import { CurrencyData } from '../../../../../lib/backend/api';

export type ClickCallback = (currency: CurrencyData) => void;

export type AvailableCurrency =
  | {
    availableCurrency?: Omit<StructuredCurrency, 'fiat'> & {
      all?: CurrencyData[];
    };
    type: 'get';
  }
  | {
    availableCurrency?: StructuredCurrency & {
      all?: CurrencyData[];
    };
    type: 'send';
  };

export type NormalDropdownProps = AvailableCurrency & {
  loading?: boolean;
  selectedCurrency: CurrencyData | null;
  searchContext: string;
  fiat?: boolean;
  onClick?: ClickCallback;
  components?: {
    NotFound?: React.FC<React.PropsWithChildren>;
    Group?: React.FC<React.PropsWithChildren>;
    Item?: React.FC<
      React.PropsWithChildren<
        | { loading: true }
        | {
          loading?: false;
          currency: CurrencyData;
          selected?: boolean;
          onClick?: () => void;
        }
      >
    >;
  };
  groupHeight?: number;
  itemHeight?: number;
  maxHeight?: number;
  embed?: boolean;
};
