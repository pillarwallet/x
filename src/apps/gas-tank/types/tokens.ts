export type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
};

export type PayingToken = {
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalUsd: number;
  totalRaw: string;
  chainId: number;
  address: string;
};

export enum SearchType {
  MyHoldings = 'My Holdings',
}

export enum SortType {
  Up = 'Up',
  Down = 'Down',
}
