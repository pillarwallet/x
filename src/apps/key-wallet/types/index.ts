export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

