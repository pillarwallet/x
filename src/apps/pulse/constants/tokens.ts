export const STABLE_CURRENCIES = [
  { chainId: 10, address: '0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85' }, // USDC on Optimism
  { chainId: 137, address: '0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359' }, // USDC on Polygon
  { chainId: 8453, address: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913' }, // USDC on Base
  { chainId: 42161, address: '0xaf88d065e77c8cC2239327C5EDb3A432268e5831' }, // USDC on Arbitrum
  { chainId: 56, address: '0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d' }, // USDC on BNB Smart Chain
  { chainId: 100, address: '0x2a22f9c3b484c3629090FeED35F17Ff8F88f76F0' }, // USDC on Gnosis
];
