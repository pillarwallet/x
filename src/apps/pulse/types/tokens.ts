type ParsedToken = {
  name: string;
  symbol: string;
  markrtCap: number;
  liquidity: number;
  volume: number;
  logo: string;
  price: number;
  priceChange: number;
  blockchain: string;
}

type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
}