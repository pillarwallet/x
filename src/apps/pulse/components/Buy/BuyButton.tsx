import { ExpressIntentResponse } from '@etherspot/intent-sdk/dist/cjs/sdk/types/user-intent-types';
import { TailSpin } from 'react-loader-spinner';
import { PayingToken, SelectedToken } from '../../types/tokens';
import { getChainName } from '../../utils/constants';

// components
import HighDecimalsFormatted from '../../../pillarx-app/components/HighDecimalsFormatted/HighDecimalsFormatted';

// utils
import { limitDigitsNumber } from '../../../../utils/number';

function getButtonText(
  isLoading: boolean,
  isInstalling: boolean,
  isFetching: boolean,
  areModulesInstalled: boolean | undefined,
  selectedToken: SelectedToken | null,
  debouncedUsdAmount: string,
  payingToken?: PayingToken,
  isDisabled?: boolean
) {
  if (areModulesInstalled === false && payingToken && !isInstalling) {
    return (
      <div className="flex text-sm items-center justify-center">{`Enable Trading on ${getChainName(payingToken.chainId)}`}</div>
    );
  }

  if (isLoading || isInstalling || isFetching) {
    return (
      <div className="flex items-center justify-center">
        <TailSpin color="#FFFFFF" height={15} width={15} />
      </div>
    );
  }

  const usdAmount = parseFloat(debouncedUsdAmount);
  const tokenUsdValue = selectedToken?.usdValue
    ? Number(selectedToken.usdValue)
    : 0;

  if (
    !isDisabled &&
    !Number.isNaN(usdAmount) &&
    usdAmount > 0 &&
    tokenUsdValue > 0
  ) {
    const tokenAmount = usdAmount / tokenUsdValue;
    const limitedUsdAmount = limitDigitsNumber(usdAmount);
    const limitedTokenAmount = limitDigitsNumber(tokenAmount);

    return (
      <div className="flex items-center justify-center gap-1 text-base">
        <span>{selectedToken ? 'Buy' : 'Select token'}</span>
        <HighDecimalsFormatted
          value={limitedTokenAmount}
          styleNumber="text-white text-base"
          styleZeros="text-white/70 text-sm"
        />
        <span>{selectedToken?.symbol}</span>
        <span>for</span>
        <HighDecimalsFormatted
          value={limitedUsdAmount}
          moneySymbol="$"
          styleNumber="text-white text-base"
          styleZeros="text-white/70 text-sm"
        />
      </div>
    );
  }

  // eslint-disable-next-line no-nested-ternary
  return selectedToken?.symbol
    ? `Buy ${selectedToken.symbol}`
    : selectedToken === null
      ? 'Select token'
      : 'Buy';
}

export interface BuyButtonProps {
  isLoading: boolean;
  isInstalling: boolean;
  isFetching: boolean;
  areModulesInstalled: boolean | undefined;
  token: SelectedToken | null;
  debouncedUsdAmount: string;
  payingTokens: PayingToken[];
  handleBuySubmit: () => Promise<void>;
  expressIntentResponse: ExpressIntentResponse | null | { error: string };
  usdAmount: string;
  notEnoughLiquidity: boolean;
}

export default function BuyButton(props: BuyButtonProps) {
  const {
    areModulesInstalled,
    debouncedUsdAmount,
    expressIntentResponse,
    handleBuySubmit,
    isFetching,
    isInstalling,
    isLoading,
    payingTokens,
    token,
    usdAmount,
    notEnoughLiquidity,
  } = props;

  const isDisabled = () => {
    if (isInstalling || isFetching) {
      return true;
    }
    if (notEnoughLiquidity) {
      return true;
    }
    if (!areModulesInstalled && payingTokens.length > 0) {
      return false;
    }
    return (
      isLoading ||
      !token ||
      !(parseFloat(usdAmount) > 0) ||
      !expressIntentResponse ||
      !!(expressIntentResponse as { error: string }).error ||
      (expressIntentResponse as ExpressIntentResponse)?.bids?.length === 0
    );
  };

  return (
    <button
      className="flex-1 items-center justify-center"
      onClick={handleBuySubmit}
      disabled={isDisabled()}
      style={{
        backgroundColor: isDisabled() ? '#29292F' : '#8A77FF',
        color: isDisabled() ? 'grey' : '#FFFFFF',
        borderRadius: 8,
      }}
      type="button"
      data-testid="pulse-buy-button"
    >
      <span className="text-base font-medium tracking-tight leading-none text-center">
        {getButtonText(
          isLoading,
          isInstalling,
          isFetching,
          areModulesInstalled,
          token,
          debouncedUsdAmount,
          payingTokens.length > 0 ? payingTokens[0] : undefined,
          isDisabled()
        )}
      </span>
    </button>
  );
}
