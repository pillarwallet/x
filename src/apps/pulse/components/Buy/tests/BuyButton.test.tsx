/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { ExpressIntentResponse } from '@etherspot/intent-sdk/dist/cjs/sdk/types/user-intent-types';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// types
import { PayingToken, SelectedToken } from '../../../types/tokens';

// components
import BuyButton from '../BuyButton';

const mockToken: SelectedToken = {
  name: 'Test Token',
  symbol: 'TEST',
  logo: 'test-logo.png',
  usdValue: '100.00',
  dailyPriceChange: 0.05,
  chainId: 1,
  decimals: 18,
  address: '0x1234567890123456789012345678901234567890',
};

const mockPayingToken: PayingToken = {
  name: 'USD Coin',
  symbol: 'USDC',
  logo: 'usdc-logo.png',
  actualBal: '100.00',
  totalUsd: 100.0,
  totalRaw: '100000000',
  chainId: 1,
  address: '0xA0b86a33E6441b8C4C8C0C4C8C0C4C8C0C4C8C0C4',
};

const mockExpressIntentResponse: ExpressIntentResponse = {
  intentHash: '0xIntentHash123456789',
  bids: [{ bidHash: '0xBidHash123456789' }],
} as any;

const mockProps = {
  isLoading: false,
  isInstalling: false,
  isFetching: false,
  areModulesInstalled: true,
  token: mockToken,
  debouncedUsdAmount: '100.00',
  payingTokens: [mockPayingToken],
  handleBuySubmit: vi.fn(),
  expressIntentResponse: mockExpressIntentResponse,
  usdAmount: '100.00',
  notEnoughLiquidity: false,
};

describe('<BuyButton />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer.create(<BuyButton {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders correct text', () => {
    it('with token and valid amount', () => {
      render(<BuyButton {...mockProps} />);

      expect(screen.getByText('Buy')).toBeInTheDocument();
      expect(screen.getByText('1')).toBeInTheDocument();
      expect(screen.getByText('TEST')).toBeInTheDocument();
      expect(screen.getByText('for')).toBeInTheDocument();
      expect(screen.getByText('$100')).toBeInTheDocument();
    });

    it('shows "Buy" when no token selected', () => {
      render(
        <BuyButton {...mockProps} token={null} debouncedUsdAmount="100.00" />
      );

      expect(screen.getByRole('button')).toHaveTextContent('Buy');
    });

    it('shows "Buy" when no amount entered', () => {
      render(<BuyButton {...mockProps} debouncedUsdAmount="0" usdAmount="0" />);

      expect(screen.getByRole('button')).toHaveTextContent('Buy');
    });

    it('shows loading spinner when loading', () => {
      render(<BuyButton {...mockProps} isLoading />);

      expect(screen.getByRole('button')).toHaveTextContent('');
    });

    it('shows loading spinner when installing', () => {
      render(<BuyButton {...mockProps} isInstalling />);

      expect(screen.getByRole('button')).toHaveTextContent('');
    });

    it('shows loading spinner when fetching', () => {
      render(<BuyButton {...mockProps} isFetching />);

      expect(screen.getByRole('button')).toHaveTextContent('');
    });

    it('shows enable trading text when modules not installed', () => {
      render(
        <BuyButton
          {...mockProps}
          areModulesInstalled={false}
          payingTokens={[mockPayingToken]}
        />
      );

      expect(
        screen.getByText('Enable Trading on Ethereum')
      ).toBeInTheDocument();
    });
  });

  describe('button states', () => {
    it('is enabled when all conditions are met', () => {
      render(<BuyButton {...mockProps} />);

      const button = screen.getByRole('button');
      expect(button).not.toBeDisabled();
      expect(button).toHaveStyle({
        backgroundColor: '#8A77FF',
        color: '#FFFFFF',
      });
    });

    it('is disabled when loading', () => {
      render(<BuyButton {...mockProps} isLoading />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
      expect(button).toHaveStyle({
        backgroundColor: '#29292F',
        color: 'grey',
      });
    });

    it('is disabled when installing', () => {
      render(<BuyButton {...mockProps} isInstalling />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when fetching', () => {
      render(<BuyButton {...mockProps} isFetching />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when no token selected', () => {
      render(<BuyButton {...mockProps} token={null} />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when no amount entered', () => {
      render(<BuyButton {...mockProps} usdAmount="0" debouncedUsdAmount="0" />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when no express intent response', () => {
      render(<BuyButton {...mockProps} expressIntentResponse={null} />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when express intent has error', () => {
      render(
        <BuyButton
          {...mockProps}
          expressIntentResponse={{ error: 'Intent failed' } as any}
        />
      );

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when no bids available', () => {
      render(
        <BuyButton
          {...mockProps}
          expressIntentResponse={{
            ...mockExpressIntentResponse,
            bids: [],
          }}
        />
      );

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is disabled when not enough liquidity', () => {
      render(<BuyButton {...mockProps} notEnoughLiquidity />);

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('is enabled when modules not installed but paying tokens available', () => {
      render(
        <BuyButton
          {...mockProps}
          areModulesInstalled={false}
          payingTokens={[mockPayingToken]}
        />
      );

      const button = screen.getByRole('button');
      expect(button).not.toBeDisabled();
    });
  });

  it('calls handleBuySubmit when clicked', () => {
    render(<BuyButton {...mockProps} />);

    screen.getByRole('button').click();

    expect(mockProps.handleBuySubmit).toHaveBeenCalled();
  });

  describe('handles edge cases', () => {
    it('handles zero token USD value', () => {
      const tokenWithZeroValue = { ...mockToken, usdValue: '0' };
      render(<BuyButton {...mockProps} token={tokenWithZeroValue} />);

      expect(screen.getByText('Buy TEST')).toBeInTheDocument();
    });

    it('handles invalid USD amount', () => {
      render(
        <BuyButton
          {...mockProps}
          debouncedUsdAmount="invalid"
          usdAmount="invalid"
        />
      );

      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('handles empty paying tokens array', () => {
      render(
        <BuyButton
          {...mockProps}
          payingTokens={[]}
          areModulesInstalled={false}
        />
      );

      const button = screen.getByRole('button');
      expect(button).not.toBeDisabled();
    });
  });
});
