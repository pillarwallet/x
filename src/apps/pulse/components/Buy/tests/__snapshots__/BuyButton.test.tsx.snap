// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<BuyButton /> > renders correctly and matches snapshot 1`] = `
<button
  className="flex-1 items-center justify-center"
  data-testid="pulse-buy-button"
  disabled={false}
  onClick={[MockFunction spy]}
  style={
    {
      "backgroundColor": "#8A77FF",
      "borderRadius": 8,
      "color": "#FFFFFF",
    }
  }
  type="button"
>
  <div
    className="flex items-center justify-center gap-1"
  >
    <span>
      Buy
    </span>
    <p
      className="text-white"
    >
      1
    </p>
    <span>
      TEST
    </span>
    <span>
      for
    </span>
    <p
      className="text-white"
    >
      $
      100
    </p>
  </div>
</button>
`;
